### PR TITLE
New Addon: `player-thumb`

### LINKED ISSUE #589: Persistent thumbnail
**Describe the new addon**

Shows the project's thumbnail in the player area of the project page until the project is started, helping with the inconsistency of Scratch thumbnails showing.

### LINKED ISSUE #6461: Replace project loading screen with the projects thumbnail
### Why this would be helpful

simple and small addon that makes the loading screen a little more interesting.
the thumbnail is usually rarely seen, only when first seeing the project on the user page and front page , this lets the user see the thumbnail a little more

### How this addon works

this addon would move the current loading screen into the corner and fill the extra space with the projects thumbnail

### Possible alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Samq64**
Resolves #589
Resolves #6461

### Changes

Always shows the thumbnail under the green flag overlay. There's also now an option to use the thumbnail as the loading screen background.

### Reason for changes

It fixes some inconsistencies and is better than the gray rectangle that sometimes appears. Some users would also like to have something more interesting as their loading screen.

### Tests

Tested on Chromium 115.

### CODE REVIEW

File: addons/persistent-thumb/addon.json
Comment (lisa-wolfgang) [inline addons/persistent-thumb/addon.json:3]:
```suggestion
  "description": "Fixes a bug in the Scratch project player where thumbnails don't always load before a project is started.",
```

Comment (mxmou) [reply addons/persistent-thumb/addon.json:3]: It isn't a bug. What you see before the project starts isn't a thumbnail, it's the state of the project when it was saved. That's why it's blank for pen projects - Scratch 3 doesn't save the pen layer.

Comment (Norbiros) [reply addons/persistent-thumb/addon.json:3]: We can remove part about bug, but I think this part `Scratch project player where thumbnails don't always load ` should stay.

Comment (Norbiros) [reply addons/persistent-thumb/addon.json:3]: When I was testing this addon I didnt undetstood how it works, until mxmou told about pen extension,  so maybe we can add info about it? Maybe
`This addon always shows the project thumbnail before a project is started (even if project uses pen extension).`

**Comment by apple502j** [comment]
Could you rename the addon ID? persistent-thumb is already used im different context.

File: addons/player-thumb/addon.json
Comment (apple502j) [inline addons/player-thumb/addon.json:25]:
```suggestion
  "versionAdded": "1.28.0",
```

File: addons/player-thumb/userscript.js
Comment (apple502j) [inline addons/player-thumb/userscript.js:2]:
```suggestion
  const el = await addon.tab.waitForElement('div[class*="stage_green-flag-overlay-wrapper_"]', {
```

**Review by apple502j — APPROVED** [review]

**Review by WorldLanguages — DISMISSED** [review]
This addon is hard to explain, so I'm blocking merge until I can take a deeper look at the addon name and description.

**Comment by CST1229** [comment]
> This addon is hard to explain, so I'm blocking merge until I can take a deeper look at the addon name and description.

I suggest something like:
"Show project thumbnail under green flag overlay"
"Shows the project's actual thumbnail instead of its initial state before starting the project."

**Comment by Samq64** [comment]
Sounds a bit long, but I guess it'll work.

File: addons/player-thumb/addon.json
Comment (lisa-wolfgang) [inline addons/player-thumb/addon.json:26]:
```suggestion
  "tags": ["community", "projectPage"],
```
In my opinion, this is more of a behavior change than a theme. There's only one thing that's having its appearance changed, and the content is dynamic.

File: addons/player-thumb/addon.json
Comment (lisa-wolfgang) [inline addons/player-thumb/addon.json:25]: Reminder to change this before merge

**Review by lisa-wolfgang — CHANGES_REQUESTED** [review]

File: addons/player-thumb/userscript.js
Comment (lisa-wolfgang) [inline addons/player-thumb/userscript.js:7]: I think compatibility with the "high resolution thumbnails" addon should be considered.

File: addons/player-thumb/userstyle.css
Comment (lisa-wolfgang) [inline addons/player-thumb/userstyle.css:6]: I'm not sure if it's within the scope of the addon to remove the tint from the green flag overlay.

File: addons/player-thumb/userscript.js
Comment (Samq64) [reply addons/player-thumb/userscript.js:7]: I don't think thumbnails are stored at higher resolutions that, but I could be wrong.

Comment (BroJac5246) [reply addons/player-thumb/userscript.js:7]: Does http://uploads.scratch.mit.edu/get_image/project/1_960x720.png (for example) work if the image is a higher resolution then 480x360?

Comment (Samq64) [reply addons/player-thumb/userscript.js]: Yes it does.

Comment (Samq64) [reply addons/player-thumb/userscript.js]: I'm always using the highest resolution regardless if high resolution thumbnails is enabled or not now.

**Comment by mxmou** [comment]
> Everything works except the thumbnail isn't darkened anymore.

That could be fixed by creating an `img` element and inserting it into the DOM between the stage and the green flag overlay. Making it disappear when the project is started shouldn't be too difficult - I think you can listen for the Redux action `scratch-gui/vm-status/SET_STARTED_STATE`. I think this change would also make the addon compatible with `hi-res-thumbnails`.

**Comment by Samq64** [comment]
@mxmou I have a z-index issue with the img element. Could you help me with it?

**Comment by Samq64** [comment]
Thanks @mxmou!

File: addons/player-thumb/userstyle.css
Comment (mxmou) [inline addons/player-thumb/userstyle.css:8]: Subtracting 1 pixel isn't enough - the thumbnail still covers the bottom and right boder.
```suggestion
  width: calc(100% - 2px);
  height: calc(100% - 2px);
```

**Review by mxmou — CHANGES_REQUESTED** [review]

**Review by mxmou — APPROVED** [review]

**Review by WorldLanguages — CHANGES_REQUESTED** [review]
Bug:
Go to https://scratch.mit.edu/create (while logged in)
Broken image with URL `https://uploads.scratch.mit.edu/get_image/project/editor_480x360.png` will be added above the stage, inside the editor.

**Comment by Samq64** [comment]
> Bug: Go to https://scratch.mit.edu/create (while logged in) Broken image with URL `https://uploads.scratch.mit.edu/get_image/project/editor_480x360.png` will be added above the stage, inside the editor.

I can't reproduce on Chromium.

**Comment by WorldLanguages** [comment]
Related discussion: #6461

File: addons/player-thumb/addon.json
Comment (mxmou) [inline addons/player-thumb/addon.json:37]: The description should mention this option, especially if it's enabled by default.

File: addons/player-thumb/userscript.js
Comment (mxmou) [inline addons/player-thumb/userscript.js:22]:
```suggestion
```

File: addons/player-thumb/userscript.js
Comment (mxmou) [inline addons/player-thumb/userscript.js:33]:
```suggestion
```

File: addons/player-thumb/userstyle.css
Comment (mxmou) [inline addons/player-thumb/userstyle.css:10]: The thumbnail is stretched while the loading screen is visible and it changes size when the project loads. This could be fixed by adding a "loading" class and some CSS:
```suggestion
}

.sa-project-thumb-loading {
  top: calc(2.75rem + 1px);
  height: calc(100% - 2.75rem - 2px);
}
```

**Review by mxmou — CHANGES_REQUESTED** [review]

**Comment by Joeclinton1** [comment]
I wonder if once this addon is made you could take it a step further, by having the ability to put custom text in the loading screen, which could be stored in a comment. Then you could have funny/useful messages just for scratch addon users.

**Comment by Samq64** [comment]
Is there a way to remove the block animation transparency? Other than that I think this is ready.

**Review by TheColaber — CHANGES_REQUESTED** [review]
seems like it doesnt work well in full screen 
image

**Comment by TheColaber** [comment]
same issue occurs... but once the project loads it fixes itself

File: addons/player-thumb/loading.css
Comment (mxmou) [inline addons/player-thumb/loading.css:2]:
```suggestion
  background-color: rgba(0, 0, 0, 0.25);
  border-radius: calc(0.5rem - 1px);
```

File: addons/player-thumb/loading.css
Comment (mxmou) [inline addons/player-thumb/loading.css:1]: > Is there a way to remove the block animation transparency?

I think the simplest way is to use a filter:
```css
body:not(.sa-body-editor) [class*="loader_block-animation_"] {
  filter: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg'><filter id='opaque'><feComponentTransfer><feFuncA type='linear' slope='3.5'></feFuncA></feComponentTransfer></filter></svg>#opaque");
}
```

Comment (mxmou) [reply addons/player-thumb/loading.css:1]: It could be in a SVG file in the addon directory instead of a data URI.

File: addons/player-thumb/addon.json
Comment (mxmou) [inline addons/player-thumb/addon.json:40]: Needs to be updated before merging

**Review by mxmou — APPROVED** [review]

**Comment by Samq64** [comment]
Is the loading setting necessary? The toggle doesn't currently do anything outside the project page and removing it would make the addon easier to explain.

**Comment by Samq64** [comment]
I added fullscreen and editor support locally and removed the setting. Should I push the changes or wait for this to be merged first?

**Comment by mxmou** [comment]
I think it makes sense to remove the setting - as you said, it would be easier to explain the purpose of the addon. Showing the thumbnail while loading is a more visible change than showing it under the green flag overlay, so it should probably be the main feature mentioned in the name and description.

**Review by mxmou — APPROVED** [review]

File: addons/player-thumb/addon.json
Comment (mxmou) [inline addons/player-thumb/addon.json:3]: The description could probably be better - it can currently be read as
`Shows the project's thumbnail instead of its initial state (before starting the project and while it loads)`
instead of
`Shows the project's thumbnail (instead of its initial state before starting the project) and (while it loads)`

Comment (WorldLanguages) [reply addons/player-thumb/addon.json:3]:
```suggestion
  "description": "Shows the project's thumbnail before starting the project and while it loads.",
```

Comment (WorldLanguages) [reply addons/player-thumb/addon.json:3]:
```suggestion
  "description": "Shows the project's thumbnail as a placeholder before the project is started and while it loads.",
```

File: addons/player-thumb/addon.json
Comment (WorldLanguages) [inline addons/player-thumb/addon.json:3]: Leaving this comment open for a few days...

**Pull request merged**
